### PR TITLE
[NNX] NNX migration prep (6/N): NNX-native DPO

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## NNX Migration Route Map

1. ✅ Add NNX scaffolding: `pure_nnx` flag, `init_state_fn`, `TrainStateNNX`, NNX utils. Linen workflow unchanged. (PR #3427)
2. ✅ NNX sharding utilities: `get_abstract_state_nnx`, `get_named_sharding_nnx`, `set_named_sharding_nnx`, `get_partition_spec_nnx`, `get_mesh_from_config`. (PR #3470)
3. ✅ NNX fully supported end-to-end: `TrainStateNNX`, model creation, gradient accumulation, checkpointing, and training loop dispatch. (PR #3500)
4. ✅ Sharding diagnostics on NNX, plus post-training bugfixes that surfaced once the NNX path got exercised end-to-end. (PR #3652)
4.5. ✅ Linen↔NNX checkpoint converter. (PR #3843)
4.6. ❌ Linen↔NNX checkpoint comparator (sibling branch on PR4.5).
5. ✅ NNX correctness fixes, feature enablements, and vocab tiling on NNX.
6. 🔄 **[This PR]** NNX-native DPO. Adds optional `reference_model: nnx.Module` field on `TrainStateNNX` and a new `dpo_loss_fn_nnx`; removes the `NotImplementedError` from train_step + eval_step NNX branches. Linen DPO path behaviorally unchanged. Stacks on PR5.
7. ❌ NNX-native MaxEngine inference.
8. ❌ NNX-native LoRA + GRPO.
9. ❌ NNX-aware QK-Clip + remaining checkpoint utilities.
9.5. ❌ NNX + AQT in MaxEngine + serve-mode reload + gpt3 prefill fix.
10. ❌ Vocab tiling `custom_vjp` for NNX.
11. ❌ Set NNX defaults to `True`; regenerate sharding goldens; flip back integration-test `pure_nnx=False` annotations.
12. ❌ Delete Linen-specific code paths and NNX compatibility flags.

## Summary

Implements NNX-native DPO. The `pure_nnx=True` training path no longer raises `NotImplementedError` on `use_dpo` runs.

The Linen DPO overlay pattern (`model.apply(params=..., reference_params=...)`) does not translate to NNX modules, which carry their parameters internally. Instead, the policy and reference models are held as separate `nnx.Module` instances on `TrainStateNNX`, and a new `dpo_loss_fn_nnx` runs both forwards with `stop_gradient` on the reference logits.

This is the next step in the NNX migration after `feat/nnx-correctness-fixes` — it closes the remaining hard `NotImplementedError` on the NNX path and unblocks default-flipping `pure_nnx=True` for users who run DPO.

### `TrainStateNNX` ([src/maxtext/layers/train_state_nnx.py](src/maxtext/layers/train_state_nnx.py))

- Optional `reference_model: nnx.Module` field. `apply_gradients` continues to update only `self.model`, leaving `self.reference_model` bit-identical across steps.

### `dpo_utils.py` ([src/maxtext/trainers/post_train/dpo/dpo_utils.py](src/maxtext/trainers/post_train/dpo/dpo_utils.py))

- New `dpo_loss_fn_nnx(policy_model, config, data, dropout_rng, params, reference_model, is_train=True)` mirroring the Linen `dpo_loss_fn` signature so it slots into `gradient_accumulation_loss_and_grad`'s dispatcher (`dropout_rng` / `params` slots are unused for NNX; `reference_model` is the single `extra_dpo_args` entry).
- Reference forward wrapped in `jax.lax.stop_gradient`. With `nnx.value_and_grad(..., argnums=0)` over the policy, no gradient flows to the reference model's `nnx.Param` leaves.
- Both `dpo_loss_fn` (Linen) and `dpo_loss_fn_nnx` (NNX) now include `indexer_loss=0.0` / `mtp_loss=0.0` in their aux dicts so the gradient-accumulation aux pytree shape matches the non-DPO `loss_fn`.
- The Linen `_split_dpo_state` / `_merge_dpo_state` overlay pattern has no NNX counterpart by design — NNX holds the reference as a sibling field on `TrainStateNNX`, and `apply_gradients` already only touches `self.model`, so no split/merge is needed.

### `train.py` ([src/maxtext/trainers/pre_train/train.py](src/maxtext/trainers/pre_train/train.py))

- `NotImplementedError` removed from `train_step` and `eval_step` NNX branches; both dispatch to `dpo_loss_fn_nnx` when `use_dpo`, with `state.reference_model` passed as `extra_dpo_args[0]`.
- `diff_wrapper` picks `_loss_fn` / `extra_dpo_args` from the per-path init block so GA and non-GA NNX paths route DPO identically.
- `_split_dpo_state` checkpoint-save stripping is now Linen-only (NNX saves the whole `TrainStateNNX` including `reference_model`; the step-0 reload later overwrites `reference_model`).

### `train_utils.py` ([src/maxtext/utils/train_utils.py](src/maxtext/utils/train_utils.py))

- NNX `init_state_fn` materializes a frozen reference model alongside the policy when `config.use_dpo`. Both are constructed via `_create_model_partial()` with `config.init_weights_seed`, so they start identical (standard DPO practice) until the step-0 reload.
- Step-0 checkpoint reload writes `step0_state["model"]` into `state["reference_model"]`. Linen path unchanged.

### Tests

- New [tests/unit/dpo_nnx_test.py](tests/unit/dpo_nnx_test.py) (7 tests):
  - `TrainStateNNX(reference_model=...)` init / `hasattr` semantics
  - `apply_gradients` leaves the reference model bit-identical
  - `dpo_loss_fn_nnx` aux key set
  - Identical policy/reference yields `loss = log(2)` and `reward_accuracy = 0.0` (strict `>` on equal logratios)
  - `dropout_rng` / `params` slots are signature-compat only — passing arbitrary values does not change the result
  - `nnx.value_and_grad(..., argnums=0)` over the policy yields finite grads on policy params only
- [tests/unit/train_nnx_test.py](tests/unit/train_nnx_test.py): dropped two stale negative tests:
  - `vocab_tiling_raises_not_implemented` — vocab tiling on NNX was implemented in `feat/nnx-correctness-fixes`
  - `train_step_dpo_raises_for_nnx` — DPO on NNX is implemented here

## Linen invariant

Linen DPO is behaviorally unchanged: the only Linen-side change is that `dpo_loss_fn` now includes `indexer_loss=0.0` / `mtp_loss=0.0` in `aux`. Existing callers either ignore those keys or read them with `.get(..., 0.0)`, so this is a no-op for downstream code. The NNX non-DPO path is unchanged: every new code path is gated on `config.use_dpo`.

## Stats

- 4 source files + 2 test files modified, **+412 / −41 lines**
- All **27 tests pass** across `train_nnx_test`, `dpo_nnx_test` (new), `train_state_nnx_test`, `train_utils_nnx_test`, `gradient_accumulation_nnx_test`

## Followups (not in this PR)

- DPO + NNX with `gradient_accumulation_steps > 1` is wired up via the same dispatcher (`extra_dpo_args` carries `reference_model` and `gradient_accumulation_loss_and_grad`'s NNX branch already calls `grad_func(local_model, config, data, None, None, *extra_dpo_args, is_train=True)` against `dpo_loss_fn_nnx`'s signature) but is not exercised by an explicit GA-DPO test. Tracked as a followup.
- Step-0 abstract-state shape mismatch (DPO-shape abstract state used to restore a non-DPO step-0 checkpoint) is mirrored from the Linen path's existing behavior; revisit only if integration testing surfaces issues.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tests/integration/setup_train_loop_nnx_test.py [hunk 1]
```diff
@@ -126,15 +126,6 @@ def test_pure_nnx_setup_param_only_split_matches_model(self):
...
     del model
 
-  def test_pure_nnx_dpo_raises_not_implemented(self):
-    """The use_dpo branch (train_utils.py:319-320) must raise for NNX."""
-    # use_dpo requires a few prerequisites; the simplest is to set the flag and
-    # let setup_train_loop reach the NotImplementedError check before the more
-    # involved DPO path runs.
-    config = _tiny_nnx_pyconfig(use_dpo=True, packing=False)
-    with self.assertRaises(NotImplementedError):
-      setup_train_loop(config, recorder=None)
```
Comment (bvandermoon) [inline tests/integration/setup_train_loop_nnx_test.py:136]: Do we need a different training loop test in place of this?

File: tests/unit/train_nnx_test.py [hunk 2]
```diff
@@ -174,16 +174,6 @@ def test_train_step_returns_state_and_metrics(self):
...
 
-  def test_train_step_dpo_raises_for_nnx(self):
-    cfg, ts = _build_state()
-    cfg.use_dpo = True
-    state_graphdef, state_pure = nnx.split(ts)
-    data = _make_data(batch=cfg.micro_batch_size_to_train_on, vocab=cfg.vocab_size)
-    with self.assertRaises(NotImplementedError):
-      pre_train.train_step(
-          state_graphdef, cfg, state_mesh_shardings=None, params_shardings=None, state=state_pure, data=data
-      )
```
Comment (bvandermoon) [inline tests/unit/train_nnx_test.py:185]: Do we need a different train test in place of this now?

File: tests/integration/setup_train_loop_nnx_test.py (hunk 1 above)
Comment (ecnal-cienet) [reply tests/integration/setup_train_loop_nnx_test.py:136]: Yes — added `test_pure_nnx_dpo_setup_materializes_reference_model`.

File: tests/unit/train_nnx_test.py (hunk 2 above)
Comment (ecnal-cienet) [reply tests/unit/train_nnx_test.py:185]: Yes — replaced with a positive test: `TestTrainStepDPONNX::test_train_step_dpo_runs_for_nnx`

File: src/maxtext/utils/muon_utils.py [hunk 3]
```diff
@@ -154,7 +154,7 @@ def get_leaf_info(leaf):
   print("\nIs this reasonable?")
 
 
-def get_model_mdn(model_name, scan_layers=True, verbose=False, pure_nnx=False):
+def get_model_mdn(model_name, scan_layers=True, verbose=False, pure_nnx=True):
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/utils/muon_utils.py]: Done

File: src/maxtext/trainers/post_train/dpo/dpo_utils.py [hunk 4]
```diff
@@ -148,10 +150,147 @@ def dpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_t
       "total_weights": total_weights,
       "moe_lb_loss": moe_lb_loss,
       "reward_accuracy": reward_accuracy,
+      "indexer_loss": 0.0,  # for gradient_accumulation aux pytree compatibility
+      "mtp_loss": 0.0,  # for gradient_accumulation aux pytree compatibility
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/trainers/post_train/dpo/dpo_utils.py:161]: Applied the same `sum` -under-GA fix to the Linen `dpo_loss_fn` in this PR.

File: src/maxtext/trainers/post_train/dpo/dpo_utils.py [hunk 5]
```diff
@@ -148,10 +150,147 @@ def dpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_t
...
+  inputs_position = jnp.concatenate([data["chosen_position"], data["rejected_position"]], 0)
+  inputs_segmentation = jnp.concatenate([data["chosen_segmentation"], data["rejected_segmentation"]], 0)
+
+  logits = policy_model(
+      decoder_input_tokens=inputs,
+      decoder_positions=inputs_position,
+      decoder_segment_ids=inputs_segmentation,
+      enable_dropout=config.enable_dropout if is_train else False,
+  )
+  intermediate_outputs = nnx.state(policy_model, nnx.Intermediate).to_pure_dict()
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/trainers/post_train/dpo/dpo_utils.py]: Done

File: src/maxtext/trainers/post_train/dpo/dpo_utils.py [hunk 6]
```diff
@@ -148,10 +150,147 @@ def dpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_t
...
+
+  # DPO loss from chosen and rejected logratios
+  LABEL_SMOOTHING, BETA = config.dpo_label_smoothing, config.dpo_beta
+  logratios_delta = BETA * (chosen_logratios - rejected_logratios)  # [B]
+  losses = (  # [B]
+      -jax.nn.log_sigmoid(BETA * logratios_delta) * (1 - LABEL_SMOOTHING)
+      - jax.nn.log_sigmoid(-BETA * logratios_delta) * LABEL_SMOOTHING
+  )
+  total_loss, total_weights = jnp.mean(losses), losses.shape[0]
+  loss = total_loss
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/trainers/post_train/dpo/dpo_utils.py]: fixed

File: src/maxtext/trainers/pre_train/train.py [hunk 7]
```diff
@@ -705,7 +721,7 @@ def train_loop(config, recorder, state=None):
 
         step_time_delta = datetime.datetime.now() - last_step_completion
 
-        state_to_save = state if not config.use_dpo else _split_dpo_state(state)[0]
+        state_to_save = state if not (config.use_dpo and not config.pure_nnx) else _split_dpo_state(state)[0]
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/trainers/pre_train/train.py:720]: > Done

File: src/maxtext/utils/muon_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/utils/muon_utils.py]: Done

File: src/maxtext/trainers/post_train/dpo/dpo_utils.py [hunk 8]
```diff
@@ -148,10 +150,147 @@ def dpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_t
...
+  rejected_segmentation = data["rejected_segmentation"][..., 1:]
+  n_logits = logits.shape[-3] // 2  # [B, S, E] - [batch, sequence, embedding/vocab]
+  chosen_logits, rejected_logits = logits[:n_logits, :, :], logits[n_logits:, :, :]
+  chosen_ref_logits, rejected_ref_logits = ref_logits[:n_logits, :, :], ref_logits[n_logits:, :, :]
+
+  # common subsequence and padding mask
+  common_prefix_mask = jnp.cumsum(chosen_ids != rejected_ids, axis=-1) == 0  # [B, S]
+  valid_seq_mask = (chosen_segmentation != 0) & (rejected_segmentation != 0) & ~common_prefix_mask  # [B, S]
+
+  # compute logratios from the sequence-reduced observed token log-probability
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/trainers/post_train/dpo/dpo_utils.py:258]: Done

**Review by bvandermoon — APPROVED** [review]
@igorts-git is adding a new DPO implementation. Can you please take a look at this migration PR Igor? We are expecting this code to go away, but it is still being migrated to NNX to keep the migration moving

**Review by igorts-git — APPROVED** [review]
The changes look good to me. However, FYI, this code will be deleted soon. I working on replacing the DPO implementation with a Tunix-based DPO implementation (see #3668).
